docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,287 @@
+
+This design should ensure compatibility between client and server in the future as these features are implemented. Neither clients nor servers will be able to access more information than what we provide in this new proto, so we need to make sure the messages and service calls anticipate needs for the above features.
+
+## Design
+
+The following is a proto file for just the rpc portion to create segments. The read portions just need to have a stream id included, which should default to being the largest stream id.
+(When we implement versions, every upload will have a new stream, which will be given the largest stream id at that time. This will allow us to get older versions by checking the stream id.)
+
+With this design, streams will be created that contain multiple parts. There were already RPCs to finalize individual segments, but each segment was not logically associated with a part, but instead just the specific object.
+
+```protobuf
+service Metainfo {
+    rpc CreateBucket(BucketCreateRequest) returns (BucketCreateResponse);
+    rpc GetBucket(BucketGetRequest) returns (BucketGetResponse);
+    rpc DeleteBucket(BucketDeleteRequest) returns (BucketDeleteResponse);
+    rpc ListBuckets(BucketListRequest) returns (BucketListResponse);
+
+    rpc CreateObject(ObjectCreateRequest) returns (ObjectCreateResponse);
+    rpc CommitObject(ObjectCommitRequest) returns (ObjectCommitResponse);
+    rpc ListObjects(ObjectListRequest) returns (ObjectListResponse);
+
+    rpc DeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+    rpc FinishDeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+
+    rpc CreateSegment(SegmentCreateRequest) returns (SegmentCreateResponse);
+    rpc AbortSegment(SegmentAbortRequest) returns (SegmentAbortResponse);
+    rpc CommitSegment(SegmentCommitRequest) returns (SegmentCommitResponse);
+    rpc MakeInlineSegment(SegmentMakeInlineRequest) returns (SegmentMakeInlineResponse);
+
+    rpc Batch(BatchRequest) returns (BatchResponse);
+
+    rpc ListSegments(SegmentListRequest) returns (SegmentListResponse);
+    rpc DownloadSegment(SegmentDownloadRequest) returns (SegmentDownloadResponse);
+}
+
+message Bucket {
+    bytes        name = 1;
+    CipherSuite  path_cipher = 2;
+    bytes        attribution_id = 3;
+
+    google.protobuf.Timestamp created_at = 4;
+    
+    int64                default_segment_size = 5;
+    RedundancyScheme     default_redundancy_scheme = 6;
+    EncryptionParameters default_encryption_parameters = 7;
+}
+
+message BucketCreateRequest {
+    Bucket bucket = 1; // ignores created_at
+}
+
+message BucketCreateResponse {}
+
+message BucketGetRequest {
+    bytes name = 1;
+}
+
+message BucketGetResponse {
+    Bucket bucket = 1;
+}
+
+message BucketDeleteRequest {
+    bytes name = 1;
+}
+message BucketDeleteResponse {}
+
+message BucketListRequest {
+    bytes     cursor    = 1;
+    Direction direction = 2;
```

Comment:
We should eliminate Direction from the protocol entirely if nothing (miniogw) currently uses anything other than the Forward direction

Reply:
okay, Jeff searched minio, and looks like we're good to remove this.

Reply:
Note: we have these already implemented, also listv2 does use it:
```
	var firstFull Key
	if !reverse && !opts.StartAfter.IsZero() {
		firstFull = joinKey(opts.Prefix, opts.StartAfter)
	}
	if reverse && !opts.EndBefore.IsZero() {
		firstFull = joinKey(opts.Prefix, opts.EndBefore)
	}
	err = store.Iterate(ctx, IterateOptions{
		Prefix:  opts.Prefix,
		First:   firstFull,
		Reverse: reverse,
		Recurse: opts.Recursive,
	}, iterate)
```

In code it's called end before.